pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -235,33 +327,74 @@ func nodesWherePreemptionMightHelp(nodes []*framework.NodeInfo, m framework.Node
 	return potentialNodes
 }
 
+type candidateList struct {
+	idx int32
+	arr []Candidate
+}
+
+func newCandidateList(size int32) *candidateList {
+	return &candidateList{idx: -1, arr: make([]Candidate, size)}
+}
+
+// add adds a new candidate to the internal array atomically.
+func (cl *candidateList) add(c *candidate) {
+	if idx := atomic.AddInt32(&cl.idx, 1); idx < int32(len(cl.arr)) {
```

Comment:
semantically speaking `cap` not `len` is the max capacity, which is what we want here.

Reply:
not really (at least in golang), because the slice is already initialized with `len = cap` and if you try to access an index that is `> len` but `< cap`  in go, that's produces a panic.

Reply:
Since it's `make([]Candidate, N)`, it should be `len` right? because we're directly setting the array elements, not appending to an allocated array. Even if `cap` were greater than `len` hypothetically, we'd still want to check if `idx < len`, not `idx < cap` because `idx < cap` wouldn't guarantee that `arr[idx]` would work.

Reply:
Ah, I thought cap defines the upper range of the index.